Added new astropy.timeseries sub-package

astrofrog opened the pull request:
### Background

This pull request adds a new astropy.timeseries sub-package to the core package. The aim of this sub-package is to provide core classes for representing time series, using QTable as a foundation. The general design of this sub-package has been discussed in [APE9](https://github.com/astropy/astropy-APEs/pull/12) which is not yet accepted, but the hope is to try and bring this to a conclusion before the 3.2 release.

This new sub-package has been developed over the last year in a standalone package called [astropy-timeseries](https://astropy-timeseries.readthedocs.io) - and we [announced it was available for testing](https://www.google.com/search?q=astropy-dev+timeseries&oq=astropy-dev+timeseries&aqs=chrome..69i57.4853j0j7&sourceid=chrome&ie=UTF-8) on the astropy-dev list back in December.

### Overall design

Those of you who have used Table/QTable with mix-in columns will know that it's already possible to represent time series with a Table and a mix-in Time column. This PR just provides a sub-class of QTable that does this automatically so that this becomes more user-friendly - but fundamentally TimeSeries still has all the flexibility of Table/QTable.

There are two main classes here, TimeSeries and BinnedTimeSeries - there is a detailed discussion on this in [APE9](https://github.com/astropy/astropy-APEs/pull/12), but the short version is that in most cases people just have some quantity vs time and TimeSeries is sufficient for this, but there are cases where being explicit about the binning (e.g. with binned X-ray time series).

Note that we are subclassing from QTable instead of Table since QTable is the way we would have designed the table class from the start if we could, so since this is a new class we have the opportunity to introduce the 'sensible' behavior from the start in relation to columns with units being quantities.

### Documentation

You can find a preview of the documentation at http://astropy-timeseries.readthedocs.org

### Path forward

I think it would be nice to aim to merge this before the 3.2 feature freeze. I am opening this well in advance of the feature freeze to allow people to have time to play with it. I recommend trying out the standalone [astropy-timeseries](https://astropy-timeseries.readthedocs.io) package. If you would like to make improvements, you are welcome to make pull requests to https://github.com/aperiosoftware/astropy-timeseries as I have a script to easily convert that package into the branch for this pull request. None of the API is set in stone, so we can still discuss/work on it!

In any case, **feel free to start reviewing this now!**

### Future work

Here are some things that I think we should do, though preferably in a separate pull request

* Move the periodogram functionality in the core package to astropy.timeseries while keeping backward-compatible imports for now in astropy.stats, and ensure that this functionality can be used directly with the ``TimeSeries`` class. This would be good to have done in 3.2 for when we announce the new timeseries subpackage

* Develop a mix-in class for NDData that can be used to construct Time Series cubes - this came up as a separate way to represent time series in APE 9, and this is equally valid, so I think it would be good to support this. However, it would be good to first have the latest WCSLIB in astropy (https://github.com/astropy/astropy/issues/7929) so that we can represent time axes in WCS.

Fixes https://github.com/astropy/astropy/issues/2205

mhvk commented on docs/timeseries/times.rst:
```diff
@@ -0,0 +1,92 @@
+
+The |TimeDelta| values can be converted to different time units using e.g.::
+
+    >>> ts_rel.time.sec
```
For the example, I'd use  `ts_rel.time.to(u.s)` - otherwise people are just going to be confused about the lack of `hour` etc...

astrofrog replied:
Fixed!

mhvk commented:
Had a quick look at the overall documentation and am in favour of including this. I like that it is just the right level of starting to be useful but not overly detailed.  Definitely needs a warning that it is experimental!  And it would be good to have some tests with a small X-ray event list.

pllim commented:
There is an error in the doctest. The error is slightly different in the job with numpy-dev, which is allowed to fail.

taldcroft commented:
@astrofrog - just so you know this is on my radar and will give it a look this week.  Conversely, #8255 is awaiting review.  :smile:  Be aware I'll be mostly out of contact (and unable to work on PRs) from Apr 15 - 30.

taldcroft commented:
@astrofrog - some initial comments, mostly somewhat high level from the table perspective:

*Up-front vs. lazy checking of TimeSeries-ness*

The current implementation tries to ensure that a TimeSeries has the necessary columns, but this approach leads to somewhat intricate code, e.g. in `add_columns`.  Sometimes intricate code can mean the behavior for users is likewise intricate.  There is also opportunity for holes / problems, and right now you can get a malformed TimeSeries by deleting or renaming the time column.

So what about lazy checking, naming a property like `has_required_columns` and a class level attribute `required_columns`.  Then any methods that actually needs the columns can just check that.  You could even make a decorator `@check_required_columns` or `@needs_required_columns` for those methods to make it syntactically clear which methods require the columns.  The decorator would raise a consistent exception.

BTW, none of this needs to be specific to TimeSeries, so it could start with a generic base class that basically handles making sure that the required columns are always in the left-most position, in the order specified by `required_columns`.  So this would include at least `__init__` (munge column order at the end), `add_columns`, `rename_column` (if someone renames a column *to* `time` then it gets moved).

Doing this, for instance, lets you easily read a table directly as a TimeSeries and then fix the column names post-facto.  E.g.
```
In [29]: t = Table([[1,2], Time([3,4], format='cxcsec')])
In [30]: t.write('junk.ecsv')
In [31]: ts = TimeSeries.read('junk.ecsv')  # Currently fails, but instead could do:
In [32]: ts.rename_column('col2', 'time')  #  voila!
```

*Downgrade to QTable*

Related to the above, I'm not so keen on the auto-downgrade to `QTable`.  I think it is cleaner to let the user do what they want, which might include temporary operations where the table does not have the required columns.

taldcroft commented:
*TimeSeries in Time*

When I saw the docs I immediately thought about a class method like `Time.series()` that puts all that syntax into its own (independently) useful framework:
```
t = Time.series(time='2016-03-22T12:30:31', time_delta=3 * u.s, n_samples=10)
ts = TimeSeries(data, time=t)
```
The corresponding method for `BinnedTimeSeries` is not so obvious unfortunately.  One thing that might work is a `Time` subclass that always has the last (-1) dimension as 2, where those two values are the start and stop time.  Then you could easily do:
```
tb = Time.binned_series(start='2016-03-22T12:30:31',
...                     delta=3 * u.s, n_bins=10)
# Since this is already in a `binned_series` context, no need for the `time_bin_` prefix.
# I also prefer `delta` over `size` because the latter has ambiguous meaning.

ts = BinnedTimeSeries(data, time_bins=tb)  # The (start, end) repr feels like "time bins" to me
```
This puts the binning behavior into a `TimeBinned` object, which would have properties like `start`, `end`, `delta`.  The `delta` property would be computed on-the-fly.

 But I'm not sure if there were practical reasons for choosing the (start, dt) representation instead of (start, end).  And of course this is all a much bigger change, but you can almost do this out of the box:
```
In [2]: tb = Time([[1,2], [3,4]], format='cxcsec')

In [3]: tb
Out[3]: 
<Time object: scale='tt' format='cxcsec' value=[[1. 2.]
 [3. 4.]]>

In [4]: tb.format='iso'

In [6]: t = Table({'time': tb, 'vals': [5, 6]})

In [7]: t
Out[7]: 
<Table length=2>
                     time [2]                       vals
                      object                       int64
-------------------------------------------------- -----
1998-01-01 00:00:01.000 .. 1998-01-01 00:00:02.000     5
1998-01-01 00:00:03.000 .. 1998-01-01 00:00:04.000     6
```
Anyway a big point here is to try encapsulating the new behavior, and in this case it means:
- Moving the creation of the time series itself out of `Table` and into `Time`
- Simplifying the `Table` change by having only one new column `time` or `time_bins`.

taldcroft commented:
One little comment about high-level API (and I notice the APE is not yet accepted...).

I've been trying to move toward a consistent usage of putting modifiers at the end (French, not English) so that related classes and attributes are close in an alphabetized list.  So that would imply:
- `TimeSeriesBinned`
- `Time.series_binned`

Not a big deal, just a thought.

Another thought is the fundamental name `TimeSeries`, which makes me think of a pandas `Series`, not a `Table`.  But I'm not sure if the obvious `TimeTable` is more or less confusing for most users.

taldcroft commented on astropy/timeseries/binned.py:
```diff
@@ -0,0 +1,277 @@
+
+    _require_time_column = False
+
+    def __init__(self, data=None, time_bin_start=None, time_bin_end=None,
```
One point for the formalists is that this signature is not consistent with the base `Table` API.  Probably better to copy all the `Table` args in order.  Maybe for 4.0 we can make all args after `data` be keyword only?

astrofrog replied:
I agree with making all non-data arguments keyword-only - in fact, I'll do this here since this is a new class, so we might as well do it right straight away. I don't think anyone should rely on positional argument order for ``TimeSeries``.

taldcroft commented:
@astrofrog - so with apologies I'm out of time for a detailed look at the code and tests, but overall the docs, tests, and high-level view look good enough for WIP inclusion in 3.2.

taldcroft commented:
Note or the `BinnedTimeSeries` with the approach I suggested you would have `tb.time_bin.start` instead of `tb.time_bin_start`.  (Going back and forth about `time_bin` vs `time_bins`.  Whatever).

mhvk commented:
Somewhat random thought on initiating time series: with the new `__array_function__` protocol in numpy, we can very easily make `np.linspace(t_start, t_end, n)` work... (and also any other numpy function that makes sense for times).

taldcroft commented:
Nice. And presumably `np.arange(t_start, t_end, dt)`?  And of course `tstart + np.arange(n_samples) * dt` already works.  So definitely worth making that happen.

The downside is that users need to know these 3 different ways and make that conceptual extrapolation of using `Time` inside their favority numpy funcs.  There is also some advantage in a class method because e.g. `tstart` can be an ISO string instead of a `Time` object, so it makes the code simpler.

pllim commented on docs/timeseries/analysis.rst:
```diff
@@ -0,0 +1,213 @@
+    >>> from astropy import units as u
+    >>> from astropy.timeseries import TimeSeries
+    >>> ts_a = TimeSeries(time='2016-03-22T12:30:31',
+    ...                          time_delta=3 * u.s,
```
Indentation.

astrofrog replied:
Fixed

pllim commented on docs/timeseries/analysis.rst:
```diff
@@ -0,0 +1,213 @@
+    ...                          time_delta=3 * u.s,
+    ...                          data={'flux': [4, 3, 1, 2, 3] * u.mJy})
+    >>> ts_ab = vstack([ts_a, ts_b])
+    >>> ts_ab
```
Note: So, while indexing would sort by time, vstacking does not sort the result by time. Interesting...

I wonder if it is worth a mention of how to handle duplicates in vstack. You said that multiple `time` columns are not allowed (in the hstack section) because that would be confusing, but wouldn't duplicate time entries in the same column be equally confusing?

astrofrog replied:
Again this behavior is inherited from Table/QTable, not a decision here. We don't currently disallow duplicate time entries in the same column for now, as there might be use cases where that is valid (event lists?). I've added a note about the sorting/duplicates behavior.

pllim commented on docs/timeseries/analysis.rst:
```diff
@@ -0,0 +1,213 @@
+:meth:`~astropy.table.Table.sort` method, as for |Table|::
+
+    >>> ts = TimeSeries(time='2016-03-22T12:30:31',
+    ...                        time_delta=3 * u.s,
```
Indentation.

pllim commented on docs/timeseries/analysis.rst:
```diff
@@ -0,0 +1,213 @@
+    example_data = get_pkg_data_filename('timeseries/kplr010666592-2009131110544_slc.fits')
+    kepler = TimeSeries.read(example_data, format='kepler.fits')
+
+(see :ref:`timeseries-io` for more details about reading in data). We can then
```
This sentence should be folded into the narrative above because currently it looks weird in the rendered HTML because the plot directive above is a `:nofigs:`.

That said, I wonder if there is value is hiding the data loader snippet above. People don't like clicking extra link just to find out how to call those 4 lines.

eteq replied:
I'd say either this should be shown (i.e. not "nofigs") as @pllim suggests, or this paragraph about loading should be removed and something like ``the data from `:ref:`timeseries-io` are used here``

astrofrog replied:
Oh this was a mistake, I'm missing ``:include-source:``

pllim commented on docs/timeseries/analysis.rst:
```diff
@@ -0,0 +1,213 @@
+   :include-source:
+   :context:
+
+    kepler_folded = kepler.fold(period=2.2 * u.day, midpoint_epoch='2009-05-02T20:53:40')
```
As someone with no experience in exoplanet research, the values you picked here seem like magic numbers to me. Is explaining why you pick those values out of topic?

astrofrog replied:
I added a sentence about this and linked to a page showing how to do the periodogram.

pllim commented on docs/timeseries/masking.rst:
```diff
@@ -0,0 +1,52 @@
+
+Masking values is done in the same way as for |Table| objects (see
+`Masking and missing values
+<http://docs.astropy.org/en/stable/table/masking.html>`_). The easiest way
```
Why hardcode URL here instead of using Intersphinx? This can be confusing if people are looking at dev doc since it links to stable regardless.

astrofrog replied:
Because this was developed in a separate package originally :) Fixed now.

pllim commented on astropy/timeseries/binned.py:
```diff
@@ -0,0 +1,277 @@
+
+        This method uses the unified I/O infrastructure in Astropy which makes
+        it easy to define readers/writers for various classes
+        (http://docs.astropy.org/en/stable/io/unified.html). By default, this
```
Use intersphinx instead of hardcoding URL?

astrofrog replied:
For docstrings I think the full URLs make sense as the :ref:s would be of no use to readers.

pllim commented on astropy/timeseries/sampled.py:
```diff
@@ -0,0 +1,250 @@
+
+        This method uses the unified I/O infrastructure in Astropy which makes
+        it easy to define readers/writers for various classes
+        (http://docs.astropy.org/en/stable/io/unified.html). By default, this
```
Intersphinx

astrofrog replied:
As above

pllim commented on docs/timeseries/masking.rst:
```diff
@@ -0,0 +1,52 @@
+    >>> np.ma.median(ts['flux'])
+    4.0
+
+.. warning:: Note that masking does not yet work for columns that have units.
```
Should move warning to the top.

astrofrog replied:
Done

pllim commented on docs/timeseries/pandas.rst:
```diff
@@ -0,0 +1,73 @@
+
+* The time column is a |Time| object that supports very high precision
+  representation of times, and makes it easy to convert between different
+  time scales and formats (e.g. ISO 8601 timestamps, Julian Dates, and so on).
```
```suggestion
  time scales and formats (e.g., ISO 8601 timestamps, Julian Dates, and so on).
```

pllim commented on docs/timeseries/pandas.rst:
```diff
@@ -0,0 +1,73 @@
+* The time column is a |Time| object that supports very high precision
+  representation of times, and makes it easy to convert between different
+  time scales and formats (e.g. ISO 8601 timestamps, Julian Dates, and so on).
+* The data columns can include |Quantity| objects with units
```
```suggestion
* The data columns can include |Quantity| objects with units.
```

pllim commented on docs/timeseries/pandas.rst:
```diff
@@ -0,0 +1,73 @@
+  representation of times, and makes it easy to convert between different
+  time scales and formats (e.g. ISO 8601 timestamps, Julian Dates, and so on).
+* The data columns can include |Quantity| objects with units
+* The |BinnedTimeSeries| class includes variable width time bins
```
```suggestion
* The |BinnedTimeSeries| class includes variable-width time bins.
```

eteq commented on docs/timeseries/index.rst:
```diff
@@ -0,0 +1,333 @@
+
+    >>> from astropy.utils.data import get_pkg_data_filename
+    >>> filename = get_pkg_data_filename('timeseries/kplr010666592-2009131110544_slc.fits')  # doctest: +REMOTE_DATA
+
```
```suggestion

.. note::
    The light curve provided here is hand-picked for example purposes.  To get
    other Kepler light curves for science purposes using Python, see the
    `astroquery <https://astroquery.readthedocs.io>`_ affiliated package.

```

eteq commented on docs/timeseries/index.rst:
```diff
@@ -0,0 +1,333 @@
+
+   import matplotlib.pyplot as plt
+   plt.plot(ts.time.jd, ts['sap_flux'], 'k.', markersize=1)
+   plt.xlabel('Barycentric Julian Date')
```
Related to @pllim's comment: it's actually not at all obvious this time series is BJD until you say this.  I'm guessing you know that because it's that's what Kepler light curves are provided in? It might be useful to explicitly point that out somewhere.

More broadly... This raises the question: should the timeseries *themselves* store this information? Or is that too context-specific?

astrofrog replied:
So actually for the purposes of the example, it's enough to say 'Julian Date' in the labels. But I've now also added information about finding out the time scale, which you can get with ``ts.time.scale`` (in the Kepler example this returns tdb = barycentric dynamical time). Since this is stored by the time object, I'd rather not also have it in the time series object itself.

eteq commented on docs/timeseries/index.rst:
```diff
@@ -0,0 +1,333 @@
+    >>> from astropy import units as u
+    >>> from astropy.stats import BoxLeastSquares
+    >>> keep = ~np.isnan(ts['sap_flux'])  # doctest: +REMOTE_DATA
+    >>> periodogram = BoxLeastSquares(ts.time.jd[keep] * u.day,
```
This is a good argument for wanting to get these periodogram wrappers in time for release... this would look *much* cleaner if you could just do `BoxLeastSquares(ts[keep])...`

astrofrog replied:
I agree! I'll work on this in a separate PR though.

eteq commented on docs/timeseries/initializing.rst:
```diff
@@ -0,0 +1,277 @@
+
+The first type of time series that we will look at here is |TimeSeries|,
+which can be used for a time series which samples a continuous variable at
+discrete and instantaneous times. Initializing a |TimeSeries| can be done
```
(minor)
```suggestion
discrete, instantaneous times. Initializing a |TimeSeries| can be done
```
?  Seems clearer to me because the "and" made it sound like the two go together rather than being distinct properties.

eteq commented on docs/timeseries/initializing.rst:
```diff
@@ -0,0 +1,277 @@
+
+    >>> from astropy import units as u
+    >>> from astropy.timeseries import TimeSeries
+    >>> ts1 = TimeSeries(time='2016-03-22T12:30:31',
```
I find this a bit confusing because `time` is sort of "overloaded" as terminology.  What about `time0` in this usage?  That makes it clear that it's the *start* time, as opposed to the middle or something like that.

mhvk replied:
In my radio data reading classes, I've ended up with `start_time` and `stop_time` (in analogy with `slice.start` and `slice.stop`.

astrofrog replied:
I've added ``start_time`` to ``TimeSeries``. I'm going for ``time_*`` for all time-related attributes for consistency (good for any kind of alphabetical sorting)

eteq commented on docs/timeseries/initializing.rst:
```diff
@@ -0,0 +1,277 @@
+    2016-03-22T12:30:39.000               3.0     6.0
+    2016-03-22T12:30:44.000               2.0     3.0
+
+If you want to be able to miss out values when adding rows, you should make
```
```suggestion
If you want to be able to skip some values when adding rows, you should make
```

or possibly "mask out"?  "miss out" seems odd to me

eteq commented on docs/timeseries/io.rst:
```diff
@@ -0,0 +1,94 @@
+
+   from astropy.utils.data import get_pkg_data_filename
+   example_data = get_pkg_data_filename('timeseries/kplr010666592-2009131110544_slc.fits')
+
```
This would go with the identical suggestion I made in the getting started, presumably.

```suggestion

.. note::
    The light curve provided here is hand-picked for example purposes.  To get
    other Kepler light curves for science purposes using Python, see the
    `astroquery <https://astroquery.readthedocs.io>`_ affiliated package.
    
```

eteq commented on docs/timeseries/io.rst:
```diff
@@ -0,0 +1,94 @@
+<astropy.timeseries.BinnedTimeSeries.read>` for more details.
+
+Alternatively, you can read in the table using your own code then construct the
+time series object as described in :ref:`timeseries-initializing`.
```
```suggestion
time series object as described in :ref:`timeseries-initializing`, although then
you cannot write out another time series in the same format.
```

eteq commented on docs/timeseries/data_access.rst:
```diff
@@ -0,0 +1,194 @@
+    >>> from astropy.timeseries import TimeSeries
+    >>> ts = TimeSeries(time='2016-03-22T12:30:31',
+    ...                        time_delta=3 * u.s,
+    ...                        data={'flux': [1., 4., 5., 3., 2.],
```
I agree with @pllim here, as I was about to leave my own comment - definitely should have a unit for flux, to encourage such behavior since this is a `QTable`.  Simplar if `temp` is a temperature

astrofrog replied:
Fixed

eteq commented on docs/timeseries/times.rst:
```diff
@@ -0,0 +1,92 @@
+
+The |TimeDelta| values can be converted to different time units using e.g.::
+
+    >>> ts_rel.time.to('s')
```
```suggestion
    >>> ts_rel.time.to('second')
```

eteq commented on docs/timeseries/times.rst:
```diff
@@ -0,0 +1,92 @@
+          1458649840.0     2.0
+          1458649843.0     4.0
+
+Relative times
```
```suggestion

Times relative to other times
=============================
```
Not sold on this exact wording, but I found "relative times" to be a slightly awkward phrase

eteq commented on docs/timeseries/times.rst:
```diff
@@ -0,0 +1,92 @@
+          1458649843.0     4.0
+
+Relative times
+==============
```
(goes with above suggestion to get headings right)
```suggestion
```

eteq commented on docs/timeseries/analysis.rst:
```diff
@@ -0,0 +1,213 @@
+
+Time series can also be combined 'horizontally' or column-wise with other tables
+using the :func:`~astropy.table.hstack` function, though these should not be
+time series (as having multiple time columns would be confusing)::
```
I wonder if an example should be given about combining time series? I.e., show how you'd do this but drop the second time column or something.  I'm certain there are use cases where someone would want to combine/match time series' after all.  E.g., the `simple_downsample` applied to a more finely sampled time series and then re-combining it with a less-sampled one.

Alternatively this could be punted for now and put on the feature list as "tool to combine time series" that would check for equality an the like...

astrofrog replied:
I think what you are asking for is basically the ability to use ``join`` with time series objects since that does the kind of combination you are talking about - except that ``join`` doesn't work with time columns. This might not be trivial, so I'm opening an issue here: https://github.com/astropy/astropy/issues/8586

eteq commented on astropy/timeseries/downsample.py:
```diff
@@ -0,0 +1,136 @@
+        return np.array(result)
+
+
+def simple_downsample(time_series, time_bin_size, func=None, time_bin_start=None, n_bins=None):
```
I am not a fan of the name `simple_downsample`, as I think it's not really useful to call anything "simple" (both because "simple" is often surprisingly complex, and "simple" isn't really that descriptive).  

What about `binned_downsample`?  Or `uniform_downsample`? (depending on what you think about my next comment...)

astrofrog replied:
What about ``aggregate_downsample`` since it's basically just applying aggregate functions to each bin?

astrofrog replied:
Renamed to ``aggregate_downsample`` following Slack discussion

pllim replied:
I'll save my suggestion to rename to `agitate_downsample` for next Apr. 😉

eteq commented on astropy/timeseries/downsample.py:
```diff
@@ -0,0 +1,136 @@
+    ----------
+    time_series : :class:`~astropy.timeseries.TimeSeries`
+        The time series to downsample.
+    time_bin_size : `~astropy.units.Quantity`
```
I think there should be a way to call this that lets the user pass in a custom binning scheme.  That would mean this argument should instead be `time_bin`, which if not a scalar, would mean "use these bins" (and would raise an exception if `time_bin_start` or `n_bins` are not None.

Otherwise, `func` really ought to be the *second* argument.  It's confusing that func is in the middle of the argument list even though `time_bin_start` and `n_bins` are more related to the bin size than the `func`.

astrofrog replied:
I agree with this, however I think we can punt on it for now and add support for custom binning later. I've re-ordered the arguments to have a more natural order and made all the arguments except the first keyword-only to make sure we don't run into API issues in future though.

astrofrog replied:
Issue here: https://github.com/astropy/astropy/issues/8584

eteq commented on astropy/timeseries/downsample.py:
```diff
@@ -0,0 +1,136 @@
+        The time series to downsample.
+    time_bin_size : `~astropy.units.Quantity`
+        The time interval for the binned time series
+    func : callable, optional
```
should instead be `combine_func` or possibly `bin_combine_func` or similar.  Just `func` doesn't give enough context about what the function is for.

astrofrog replied:
Fixed

eteq commented on astropy/timeseries/downsample.py:
```diff
@@ -0,0 +1,136 @@
+def simple_downsample(time_series, time_bin_size, func=None, time_bin_start=None, n_bins=None):
+    """
+    Downsample a time series by binning values into bins with a fixed size,
+    using a single function
```
```suggestion
    using a single function to combine the values in the bin.
```

astrofrog commented on astropy/table/table.py:
```diff
@@ -830,7 +831,7 @@ def _new_from_slice(self, slice_):
     def _make_table_from_cols(table, cols, verify=True, names=None):
         """
-        Make ``table`` in-place so that it represents the given list of ``cols``.
+        Make ``~astropy.table.Table`` in-place so that it represents the given list of ``cols``.
```
This was due to a search and replace - fixed

astrofrog commented on astropy/timeseries/io/kepler.py:
```diff
@@ -0,0 +1,90 @@
+    """
+    hdulist = fits.open(filename)
+    # Get the lightcurve HDU
+    telescop = hdulist[0].header['telescop'].lower()
```
Fixed

astrofrog commented on astropy/timeseries/sampled.py:
```diff
@@ -0,0 +1,250 @@
+
+        else:
+
+            if len(self.colnames) > 0 and len(time) != len(self):
```
Fixed

astrofrog commented on astropy/timeseries/sampled.py:
```diff
@@ -0,0 +1,250 @@
+__all__ = ['TimeSeries']
+
+
+class TimeSeries(BaseTimeSeries):
```
Fixed

astrofrog commented on astropy/timeseries/sampled.py:
```diff
@@ -0,0 +1,250 @@
+                return out
+        return super().__getitem__(item)
+
+    def add_columns(self, *args, **kwargs):
```
This should be inherited from QTable, though I've now added InheritDocstrings to make sure this happens, and have added a comment in this method

astrofrog commented on astropy/timeseries/tests/test_binned.py:
```diff
@@ -0,0 +1,280 @@
+from astropy.timeseries.binned import BinnedTimeSeries
+
+
+CSV_FILE = os.path.join(os.path.dirname(__file__), 'data', 'binned.csv')
```
Fixed

astrofrog commented on astropy/timeseries/tests/test_downsample.py:
```diff
@@ -0,0 +1,69 @@
+ts_units = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4] * u.count], names=['a'])
+
+
+def test_reduceat():
```
Yeah basically only some Numpy functions have a reduceat method, so this is a way to make sure that it works for any arbitrary method. I agree it could live in astropy.utils but for now I'd rather it wasn't part of the public API until it's very stable, so I'll leave it where it is, and maybe after this PR is merged we can decided whether we want to expose it in astropy.utils.

mhvk replied:
`reduceat` is a standard method on every numpy ufunc. It doesn't really make sense on `np.sum` since that is already equal to `np.add.reduce` (like `np.min` is equal to `np.minimum.reduce`).

But for other "reductions" like `np.median` there is indeed no analogue; anyway, bottom line is that I think it needs a bit of thought of how to expose it, so best to leave it here for the moment. But could someone open an issue so we don't forget?

astrofrog replied:
Issue opened: https://github.com/astropy/astropy/issues/8585

astrofrog commented on docs/timeseries/index.rst:
```diff
@@ -0,0 +1,333 @@
+be interested in the |BinnedTimeSeries| sub-class which we show in
+:ref:`using-timeseries`).
+
+To start off, we retrieve a FITS file containing a Kepler light curve for a
```
Added a couple of links

astrofrog commented on docs/timeseries/initializing.rst:
```diff
@@ -0,0 +1,277 @@
+    >>> from astropy import units as u
+    >>> from astropy.timeseries import TimeSeries
+    >>> ts1 = TimeSeries(time='2016-03-22T12:30:31',
+    ...                         time_delta=3 * u.s,
```
Fixed

astrofrog commented on docs/timeseries/initializing.rst:
```diff
@@ -0,0 +1,277 @@
+pass multiple times to the ``time`` argument::
+
+    >>> ts2 = TimeSeries(time=['2016-03-22T12:30:31',
+    ...                               '2016-03-22T12:30:38',
```
Fixed

astrofrog commented on docs/timeseries/initializing.rst:
```diff
@@ -0,0 +1,277 @@
+---------------------------
+
+To create a binned time series with equal-size contiguous bins, it is sufficient
+to specify a start time as well as a bin size::
```
I've now clarified the exact behavior in the class docstring.

astrofrog commented on docs/timeseries/initializing.rst:
```diff
@@ -0,0 +1,277 @@
+
+
+    >>> ts5 = BinnedTimeSeries(time_bin_start=['2016-03-22T12:30:31',
+    ...                                    '2016-03-22T12:30:34',
```
Fixed

astrofrog commented on docs/timeseries/initializing.rst:
```diff
@@ -0,0 +1,277 @@
+specify an array of start times and bin widths::
+
+    >>> ts6 = BinnedTimeSeries(time_bin_start=['2016-03-22T12:30:31',
+    ...                                    '2016-03-22T12:30:38',
```
Fixed

astrofrog commented on docs/timeseries/initializing.rst:
```diff
@@ -0,0 +1,277 @@
+``time_bin_start`` and ``time_bin_end``::
+
+    >>> ts7 = BinnedTimeSeries(time_bin_start=['2016-03-22T12:30:31',
+    ...                                    '2016-03-22T12:30:33',
```
Fixed

astrofrog commented on docs/timeseries/initializing.rst:
```diff
@@ -0,0 +1,277 @@
+    2016-03-22T12:30:33.000                2.0
+    2016-03-22T12:30:40.000                1.0
+
+You can also specify vector |Time| objects directly.
```
I've removed this - this is true for all examples and it's explained in the first example on the page.

astrofrog commented on docs/timeseries/initializing.rst:
```diff
@@ -0,0 +1,277 @@
+|Table|, e.g.::
+
+    >>> ts8 = BinnedTimeSeries(time_bin_start=['2016-03-22T12:30:31',
+    ...                                    '2016-03-22T12:30:34',
```
Fixed

astrofrog commented on docs/timeseries/io.rst:
```diff
@@ -0,0 +1,94 @@
+
+
+See the documentation for :meth:`TimeSeries.read
+<astropy.timeseries.TimeSeries.read>` and :meth:`BinnedTimeSeries.read
```
I think this is fixed with the latest version of sphinx or something, because the table renders correctly for me.

astrofrog commented on docs/timeseries/data_access.rst:
```diff
@@ -0,0 +1,194 @@
+    >>> from astropy import units as u
+    >>> from astropy.timeseries import TimeSeries
+    >>> ts = TimeSeries(time='2016-03-22T12:30:31',
+    ...                        time_delta=3 * u.s,
```
Fixed

astrofrog commented on docs/timeseries/data_access.rst:
```diff
@@ -0,0 +1,194 @@
+access notation, as shown in `Accessing data`_, but they can also be accessed
+more conveniently using the |time_attr| attribute::
+
+    >>> ts.time
```
Time is a special attribute here though, time series can be of anything, not just flux, so .flux is much harder to justify.

astrofrog commented on docs/timeseries/data_access.rst:
```diff
@@ -0,0 +1,194 @@
+Extracting a subset of columns
+==============================
+
+We can create a new time series with just the ``flux`` column by doing::
```
Copy - have clarified

astrofrog commented on docs/timeseries/times.rst:
```diff
@@ -0,0 +1,92 @@
+    >>> from astropy import units as u
+    >>> from astropy.timeseries import TimeSeries
+    >>> ts = TimeSeries(time='2016-03-22T12:30:31', time_delta=3 * u.s,
+    ...                        data={'flux': [1., 3., 4., 2., 4.]})
```
Fixed

astrofrog commented on docs/timeseries/io.rst:
```diff
@@ -0,0 +1,94 @@
+series readers, these methods will try and use some of the default :class:`~astropy.table.Table`
+readers and then require users to specify the name of the important columns.
+
+For example, if you are reading in a file called ``sampled.dat`` where the time
```
Done

astrofrog commented on docs/timeseries/data_access.rst:
```diff
@@ -0,0 +1,194 @@
+
+or within a time range::
+
+   >>> ts.loc[Time('2016-03-22T12:30:31'):Time('2016-03-22T12:30:40')]
```
The last element is inclusive - this behavior is inherited from Table/QTable and I believe it is to be consistent with pandas. I added a note to that effect. Also, the times don't have to match exactly, and I've updated the example to show that.

astrofrog commented:
Thanks everyone for your reviews! :pray: :sweat_smile: :sweat_smile: :sweat_smile:

I've done my best to try and address everything, and where I couldn't address it now I've opened new issues and linked to them. I've left a few comments as unresolved above so you can see the answer, but if you are happy with it please mark them as resolved.

---

@pllim - I've accepted all your suggestions and implemented most of your comments (most of the non-suggestion changes are in 29da87bae2723d8795008dc65f45b79d9095e34e if you want to check)

Regarding the inclusion of the Kepler/TESS readers - I think that in cases where the format is well defined and documented and versioned, it's ok to have readers/writers in the core package? (after all, for io.ascii we have e.g. the IPAC reader which is reasonably specific but well documented). What I wouldn't want to see in the core package are formats that aren't actually documented. Also since Kepler/TESS time series are quite common at the moment, having these readers might entice more users to start using this subpackage. Does that seem reasonable?

Regarding the astropy-benchmarks - yes we should add some there, but that can wait until after feature freeze. It's also too soon indeed for performance tips I think.

---

@eteq - I've accepted all your suggestions and implemented most of your comments (most of the non-suggestion changes are in c7e6200132bf497d6d78728cffc64e2e04fd8d18 if you want to check).

Regarding the periodograms and moving them over to astropy.timeseries, I'll try and do this tomorrow in a separate PR.

---

@taldcroft - thanks for your detailed comments! Here are some thoughts below:

> The current implementation tries to ensure that a TimeSeries has the necessary columns, but this approach leads to somewhat intricate code, e.g. in add_columns. Sometimes intricate code can mean the behavior for users is likewise intricate. There is also opportunity for holes / problems, and right now you can get a malformed TimeSeries by deleting or renaming the time column. So what about lazy checking, naming a property like has_required_columns and a class level attribute required_columns. Then any methods that actually needs the columns can just check that. 

I see your point, but I don't really like that idea of being able to have a TimeSeries object without a time column for quite a while until eventually e.g. .fold() or another time series-specific method is used. The issue is that since this is mostly a container class with not much functionality, many callers will actually be third-party functions which will be expecting the time column to be there (and we can't decorate them). At the same time, you are right that currently one can e.g. rename or remove the time column without any warning/error, so this is clearly not ideal. My plan is to try and investigate a little more how we could get around this, but I might run out of time for the 3.2 feature freeze. However, arguably the current behavior of being able to rename or remove the time column is a bug, so I believe this is something I could address in a bug fix release in the worst case.

> Related to the above, I'm not so keen on the auto-downgrade to QTable. I think it is cleaner to let the user do what they want, which might include temporary operations where the table does not have the required columns.

I've gone back and forth on this and I do think dropping down to QTable if the user does e.g. ``ts['flux', 'error']`` is the right behavior, since if they really wanted to keep it a time series, they could easily keep the ``time`` column. I don't think a common use case is going to be to extract two non-time columns and add a new/different time column. I propose leaving this as-is for now and re-assessing over the next six months once people start using this.

> When I saw the docs I immediately thought about a class method like Time.series() that puts all that syntax into its own (independently) useful framework.

I can also see the benefit of this, but now for users creating a time series would be a two-step operation instead of a one-step operation. The class method ``Time.series`` is elegant but in my view actually makes things a little less obvious for users. The current API has been discussed for a while in the APE, so I'd rather keep it the way it is currently - it might not be as elegant, but I think it'll be more intuitive for users to just deal with the TimeSeries initializer rather than doing things in two steps.

Likewise, the vector column approach for binned time series is an elegant use of the Time/Table capabilities, but I worry that it's going to also end up being quite confusing for users. It's also not clear how indexing would work in this case.

Having said that, I'm totally on board with being able to use things like ``np.linspace`` etc on ``Time`` as I can see that being very useful.

> Another thought is the fundamental name TimeSeries, which makes me think of a pandas Series, not a Table. But I'm not sure if the obvious TimeTable is more or less confusing for most users.

Even though ``TimeTable`` would be more logical, I think most astronomers would think of ``TimeSeries`` as meaning the time + all the associated data, as a table, so I think we are ok.

---

At this point, I'll track down the last failures in CI, but otherwise @pllim and @eteq feel free to take another look and leave more comments if necessary or approve if you are happy with it. I'll be working on moving the periodograms in a separate PR.

eteq commented:
A few quick follow-ons from above:

> this approach leads to somewhat intricate code

I sympathize with @taldcroft's suggestion here. But I'm not sure I'm a fan of doing it "lazy", because it feels as likely to end up inconsistent as the approach @astrofrog is using for *un-decorated* methods. Perhaps a safer thing to do is instead centralize all the checking in one method and call it on demand as @taldcroft suggests in the lazy way *as well as* at the end of the `TimeSeries` initializer? That way it seems like we have our cake and eat it too. Relatedly, I think we would *not* want the rename example as posed by @taldfroft to work, but rather:
```python
t = Table.read('junk.ecsv')
t.rename_column('col2', 'time')
ts = TimeSeries(t)
```
Because then the `TimeSeries` object is always a time series.

Relatedly I also favor `TimeSeries` over `TimeTable` purely because that's the terminology used in astro for such things. I think it's more likely to be findable on that basis. And I think the `QTable` behavior is right for the same reason: it's most consistent with what the user might think that means.

 I really like @taldcroft's idea of the ``Time.series()`` classmethod as an *alternate* way to make a TimeSeries, though. I could definitely see myself using that. That is, it would just yield a new `TimeSeries` with that one column populated (and maybe it would also get kwargs for additional columns although that feels optional. Perhaps that can be a follow-on PR though?

pllim commented:
Re: TimeSeries vs TimeTable -- I also prefer TimeSeries, because TimeTable makes me think of something I use to schedule my events for the day.

pllim commented on astropy/timeseries/io/kepler.py:
```diff
@@ -0,0 +1,93 @@
+
+    Parameters
+    ----------
+    filename: `str` or `pathlib.Path`
```
```suggestion
    filename : `str` or `pathlib.Path`
```

pllim commented on astropy/timeseries/sampled.py:
```diff
@@ -0,0 +1,298 @@
+
+    Parameters
+    ----------
+    data : numpy ndarray, dict, list, Table, or table-like object, optional
```
```suggestion
    data : numpy ndarray, dict, list, `~astropy.table.Table`, or table-like object, optional
```

pllim commented on astropy/timeseries/sampled.py:
```diff
@@ -0,0 +1,298 @@
+
+        Parameters
+        ----------
+        filename: str
```
```suggestion
        filename : str
```

pllim commented on astropy/timeseries/sampled.py:
```diff
@@ -0,0 +1,298 @@
+            File to parse.
+        format : str
+            File format specifier.
+        time_column: str, optional
```
```suggestion
        time_column : str, optional
```

pllim commented on astropy/timeseries/sampled.py:
```diff
@@ -0,0 +1,298 @@
+            File format specifier.
+        time_column: str, optional
+            The name of the time column.
+        time_format: str, optional
```
```suggestion
        time_format : str, optional
```

pllim commented on astropy/timeseries/sampled.py:
```diff
@@ -0,0 +1,298 @@
+            The name of the time column.
+        time_format: str, optional
+            The time format for the time column.
+        time_scale: str, optional
```
```suggestion
        time_scale : str, optional
```

pllim commented on docs/stability.rst:
```diff
@@ -251,6 +251,17 @@ The current planned and existing sub-packages are:
+                astropy.timeseries
+            </td>
+            <td align='center'>
+                <span class="mature"></span>
```
Is "mature" an accurate categorization here? It conflicts with "in heavy development" just below.

astrofrog replied:
woops, copy pasta :spaghetti:

astrofrog commented:
@eteq:

> I sympathize with @taldcroft's suggestion here. But I'm not sure I'm a fan of doing it "lazy", because it feels as likely to end up inconsistent as the approach @astrofrog is using for un-decorated methods. Perhaps a safer thing to do is instead centralize all the checking in one method and call it on demand as @taldcroft suggests in the lazy way as well as at the end of the TimeSeries initializer? That way it seems like we have our cake and eat it too. Relatedly, I think we would not want the rename example as posed by @taldfroft to work, but rather:

Ok so I've now cleaned up the checking significantly in https://github.com/astropy/astropy/pull/8540/commits/e7c1807cdd76aebddc24175e56d976492284e64c - I've added a mechanism to decorate any method that adds/removes columns, and I check for consistency after the method is called. There are a few corner cases where we need to relax or disable the checks - for example when using vstack, an empty time series is temporarily created, so we need to account for that. What I've written up now deals with all these corner cases and I've made sure they are well tested so I think things are a lot more robust now.

> I really like @taldcroft's idea of the Time.series() classmethod as an alternate way to make a TimeSeries, though. I could definitely see myself using that. That is, it would just yield a new TimeSeries with that one column populated (and maybe it would also get kwargs for additional columns although that feels optional. Perhaps that can be a follow-on PR though?

I definitely won't have time for this PR

---

@pllim:

> API doc for io seems incomplete -- scroll down https://28570-2081289-gh.circle-artifacts.com/0/home/circleci/project/docs/_build/html/timeseries/index.html to see what I mean.

Should be fixed now

> In https://28570-2081289-gh.circle-artifacts.com/0/home/circleci/project/docs/_build/html/timeseries/times.html#times-relative-to-other-times , I am curious if I can format the output of the delta time, say, to '.2f'. Is that possible?

No I don't think so. Maybe open an issue for astropy.time if you think this would be important, but here 'format' refers not so much to the precision but the type of string or float representation.

eteq commented:
LGTM now too, so I'm going to pull the trigger.

> I definitely won't have time for this PR

Fair enough.  I might take a pass at it but I'll make an issue now so it could always go in 4.0 if need be.
